internal/agent-api/client.go

```diff
@@ -0,0 +1,191 @@
+	nc                *nats.Conn
+	log               *slog.Logger
+	agentId           string
+	handshakeTimeout  time.Duration
+	handshakeReceived *atomic.Bool
+
+	handshakeTimedOut  HandshakeCallback
+	handshakeSucceeded HandshakeCallback
+	eventReceived      EventCallback
+	logReceived        LogCallback
+}
+
+func NewAgentClient(nc *nats.Conn,
+	handshakeTimeout time.Duration,
+	onTimedOut HandshakeCallback,
+	onSuccess HandshakeCallback,
+	onEvent EventCallback,
+	onLog LogCallback,
+	log *slog.Logger,
+) *AgentClient {
+	shake := &atomic.Bool{}
+	shake.Store(false)
+	return &AgentClient{
+		nc:                 nc,
+		handshakeTimeout:   handshakeTimeout,
+		handshakeTimedOut:  onTimedOut,
+		handshakeSucceeded: onSuccess,
+		eventReceived:      onEvent,
+		logReceived:        onLog,
+
+		log:               log,
+		handshakeReceived: shake,
+	}
+}
+
+func (a *AgentClient) Id() string {
+	return a.agentId
+}
+
+func (a *AgentClient) Start(agentId string) error {
+	a.agentId = agentId
+	_, err := a.nc.Subscribe("agentint.handshake", a.handleHandshake)
+	if err != nil {
+		return err
+	}
+
+	_, err = a.nc.Subscribe(fmt.Sprintf("agentint.%s.events.*", agentId), a.handleAgentEvent)
+	if err != nil {
+		return err
+	}
+
+	_, err = a.nc.Subscribe(fmt.Sprintf("agentint.%s.logs", agentId), a.handleAgentLog)
+	if err != nil {
+		return err
+	}
+
+	go a.awaitHandshake(agentId)
+
+	return nil
+}
+
+func (a *AgentClient) DeployWorkload(request *DeployRequest) (*DeployResponse, error) {
+	bytes, err := json.Marshal(request)
+	if err != nil {
+		return nil, err
+	}
+
+	status := a.nc.Status()
+	a.log.Debug("NATS internal connection status",
+		slog.String("agentId", a.agentId),
+		slog.String("status", status.String()))
+
+	subject := fmt.Sprintf("agentint.%s.deploy", a.agentId)
+	resp, err := a.nc.Request(subject, bytes, 1*time.Second)
+	if err != nil {
+		if errors.Is(err, os.ErrDeadlineExceeded) {
+			return nil, errors.New("timed out waiting for acknowledgement of workload deployment")
+		} else {
+			return nil, fmt.Errorf("failed to submit request for workload deployment: %s", err)
+		}
+	}
+
+	var deployResponse DeployResponse
+	err = json.Unmarshal(resp.Data, &deployResponse)
+	if err != nil {
+		a.log.Error("Failed to deserialize deployment response", slog.Any("error", err))
+		return nil, err
+	}
+	return &deployResponse, nil
+}
+
+func (a *AgentClient) Undeploy() error {
+	subject := fmt.Sprintf("agentint.%s.undeploy", a.agentId)
+	_, err := a.nc.Request(subject, []byte{}, 500*time.Millisecond) // FIXME-- allow this timeout to be configurable... 500ms is likely not enough
+	if err != nil {
+		a.log.Warn("request to undeploy workload via internal NATS connection failed",
+			slog.String("agentId", a.agentId), slog.String("error", err.Error()))
+		return err
+	}
+	return nil
+}
+
+func (a *AgentClient) awaitHandshake(agentId string) {
+	// timeoutAt := time.Now().UTC().Add(a.handshakeTimeout)
+
+	<-time.After(a.handshakeTimeout)
```

Comment:
This pattern will prevent the node from stopping in a timely manner.

The previous machine manager code allowed the node to exit without waiting for this timeout:

```
func (m *MachineManager) awaitHandshake(vmid string) {
	timeoutAt := time.Now().UTC().Add(m.handshakeTimeout)

	handshakeOk := false
	for !handshakeOk && !m.stopping() {
		if time.Now().UTC().After(timeoutAt) {
			m.log.Error("Did not receive NATS handshake from agent within timeout.", slog.String("vmid", vmid))
			if len(m.handshakes) == 0 {
				m.log.Error("First handshake failed, shutting down to avoid inconsistent behavior")
				m.cancel()
			}
			return
		}

		_, handshakeOk = m.handshakes[vmid]
		time.Sleep(time.Millisecond * agentapi.DefaultRunloopSleepTimeoutMillis)
	}
}
```

Reply:
the timeout is a tiny number. The above code didn't actually work anymore. I was constantly getting handshake failures. It didn't start working again until I used the new method.

Waiting the remainder of a timeout measured in milliseconds before a node shuts down isn't a bad thing